CONTRIBUTING.md

```diff
@@ -88,11 +94,183 @@ We recommend the following commit structure in the following order:
+
+```bash
+export PATH=~/Library/Python/3.7/bin/:$PATH
+```
+
+## Setup Development Environment
+
+Clone the `mbed-tools` GitHub repository.
+
+```bash
+git clone git@github.com:ARMmbed/mbed-tools.git
+```
+
+Set up the development environment using tox (tox will create a development environment at mbed-tools/.venv):
+
+```bash
+cd mbed-tools/
+tox -e dev
+source .venv/bin/activate
+```
+
+## Unit Tests, Code Formatting and Static Analysis
+
+After you have activated your development environment, run `tox -e py37` to run
+unit tests and static code analysis checks:
+
+```bash
+tox -e py37
+```
+
+This will run `black`, `flake8`, `mypy` and `pytest`. If you would like to run these tools individually, see below:
+
+Run unit tests:
+
+```bash
+pytest
+```
+
+Run code formatter (it will format files in place):
+
+```bash
+black .
+```
+
+Run static analysis (note that no output means all is well):
+
+```bash
+flake8
+```
+
+Perform static type check:
+
+```bash
+mypy src/mbed_tools
+```
+
+## Documenting code
+
+Inclusion of docstrings is needed in all areas of the code for Flake8 checks in the CI to pass.
+
+We use [google-style](http://google.github.io/styleguide/pyguide.html#381-docstrings) docstrings.
+
+To set up google-style docstring prompts in Pycharm:
+
+* in the menu navigate to Preferences > Tools > Python Integrated Tools
+* in the dropdown for docstring format select 'Google'
+
+For longer explanations, you can also include markdown. Markdown can also be kept in separate files in the
+`docs/user_docs` folder and included in a docstring in the relevant place using the [reST include](https://docutils.sourceforge.io/docs/ref/rst/directives.html#including-an-external-document-fragment) as follows:
+
+```python
+    .. include:: ../docs/user_docs/documentation.md
+```
+
+### Building docs locally
+
+You can do a preview build of the documentation locally by running:
+
+```bash
+generate-docs
+```
+
+This will generate the docs and output them to `local_docs`.
+This should only be a preview. Since documentation is automatically generated
+by the CI you shouldn't commit any docs html files manually. The docs will be
+generated and checked in as part of the release process.
+
+### Viewing docs generated by the CI
+
+If you want to preview the docs generated by the CI you can view them in the Azure pipeline artifacts directory for the build.
+
+Documentation only gets committed back to this repo to the `docs`
+directory during a release and this is what gets published to Github pages.
+Don't modify any of the files in this directory by hand.
+
+## Type hints
+
+Type hints should be used in the code wherever possible. Since the
+documentation shows the function signatures with the type hints
+there is no need to include additional type information in the docstrings.
+
+## Code Climate
+
+Code Climate is integrated with our GitHub flow. Failing the configured rules will yield a pull request not mergeable.
```

Comment:
This was a preexisting issue: Can we change the wording here? "Yield a pull request not mergeable" sounds silly to me.

Reply:
"result in a ..."